### PR TITLE
 MAINT: Warn if `_add_newdocs.py` is used to add docstrings to pure-python objects

### PR DESCRIPTION
**Pull request by eric-wieser**
Continuation of the maintenance part of #13876

### CODE REVIEW

File: numpy/core/function_base.py [hunk 1]
```diff
@@ -48,7 +49,7 @@ def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None,
     .. versionchanged:: 1.16.0
         Non-scalar `start` and `stop` are now supported.
 
-    Parameters
+    Parameters
```
Comment (eric-wieser) [inline numpy/core/function_base.py]: Will remove this whitespace

**Comment by eric-wieser** [comment]
Will rebase this on gh-13946

**Comment by eric-wieser** [comment]
Rebased, pushed to the wrong branch by accident so this never updated

File: numpy/core/function_base.py [hunk 2]
```diff
@@ -427,15 +428,38 @@ def geomspace(start, stop, num=50, endpoint=True, dtype=None, axis=0):
...
+        return False
+
+    if isinstance(obj, type) and obj.__flags__ & Py_TPFLAGS_HEAPTYPE:
+        return False
+
+    return True
+
+
+def _add_docstring(obj, doc, warn_on_python):
+    if warn_on_python and not _needs_add_docstring(obj):
```
Comment (charris) [inline numpy/core/function_base.py]: The tests are complaining about the stack level of the warning.

**Comment by mattip** [comment]
Thanks @eric-wieser

**Pull request merged**